### PR TITLE
DAOS-10450 control: Always set engine NUMA node on exec

### PR DESCRIPTION
**Pull request by tanabarr**
Previous behavior is that pinned_numa_node is not always passed when
executing engine which causes engine to default to NUMA-0.

In the case where server config does not specify pinned_numa_node and
control plane derives NUMA node to be 1 (from fabric_iface "ib1"
affinity), hugepages will be allocated on NUMA-1 but engine will try to
allocate DMA buffer from NUMA-0 and fail.

Correct behavior by passing pinned_numa_node commandline parameter when
starting engine unless single engine is on NUMA-0 or first_core is set,
in which case we want to avoid passing pinned_numa_node so that the
legacy core allocation algorithm is activated in the engine. Verify that
first_core and pinned_numa_node parameters are never be set together in
server config file engine section.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

**Review by NiuYawei — APPROVED** [review]

**Review by NiuYawei — COMMENTED** [review]
On DAOS engine, there is a 'legacy' mode which doesn't have numa specified, I'm not sure if control plane has the 'legacy' mode as well? Will this patch break the 'legacy' mode.

File: src/control/server/engine/config.go
Comment (kjacque) [inline src/control/server/engine/config.go:205]: Wondering - should pinned NUMA node no longer be a pointer? If it should always be set.

**Review by kjacque — APPROVED** [review]
OK with landing as is. The comment was just a thought and could be addressed in a follow-on.

File: src/control/server/engine/config.go
Comment (tanabarr) [reply src/control/server/engine/config.go:205]: yes, good point

**Comment by tanabarr** [comment]
> On DAOS engine, there is a 'legacy' mode which doesn't have numa specified, I'm not sure if control plane has the 'legacy' mode as well? Will this patch break the 'legacy' mode.

With the patch, "-p" will not be specified when `first_core` is set non-zero and in that case legacy mode will then be activated. Legacy mode is also activated when hwloc lib doesn't return nr_numa_nodes >0.

The difference in behaviour introduced by the patch is that in the normal case where `pinned_numa_node` is not specified in the server config file and the NUMA node is set based on the affinity of the fabric interface selected, this derived NUMA node index will be passed to engine with "-p" (which will not activate legacy mode).

This change solves the issue where currently if control plane derives NUMA node 1 from fabric interface, no "-p" value is set and engine affinity defaults to 0 and is out of sync with control plane and hugepage allocation affinity.

With regard to the legacy mode changes introduced by the patch, do you think there is a problem?

File: utils/config/daos_server.yml
Comment (NiuYawei) [inline utils/config/daos_server.yml:293]: Agreed. I think the "core offset" parameter doesn't make sense to NUMA mode, and I've already changed io engine in multiprovider feature branch to ignore the 'core offset' parameter for NUMA mode.

**Review by NiuYawei — APPROVED** [review]

File: utils/config/daos_server.yml
Comment (kjacque) [inline utils/config/daos_server.yml:425]: Why double commented?

Comment (tanabarr) [reply utils/config/daos_server.yml:425]: Now I've tightened up the validation, in the test that un-comments daos_server.yml it would fail if this was specified in addition to first_core being set greater than zero.

**Review by kjacque — APPROVED** [review]

File: utils/config/daos_server.yml
Comment (kjacque) [reply utils/config/daos_server.yml:425]: Okay. I'm not a big fan of changing our config file formatting to accommodate unit tests, but it's a small thing and any user will probably do the right thing anyway in the short term. With the two fields incompatible now, we should probably come up with another way of generating our test configs in the longer term.

**Comment by tanabarr** [comment]
@mjmac I would like to land this PR separately to the changes requested regarding using pmem as the canonical source of engine affinity. This PR specifically addresses a problem where  the engine is started bound to a NUMA node that has no hugepages allocated on and DMA buffer memory allocation fails as a result. I think the changes to use pmem to derive affinity should be done as a separate PR as they will require nontrivial work to retrieve pmem socket-ID from scan results and integrate into storage config. That work would cloud the scope of this PR IMO and delay this fix which is needed to resolve CI (and customer) issues.

**Comment by mjmac** [comment]
> @mjmac I would like to land this PR separately to the changes requested regarding using pmem as the canonical source of engine affinity. This PR specifically addresses a problem where the engine is started bound to a NUMA node that has no hugepages allocated on and DMA buffer memory allocation fails as a result. I think the changes to use pmem to derive affinity should be done as a separate PR as they will require nontrivial work to retrieve pmem socket-ID from scan results and integrate into storage config. That work would cloud the scope of this PR IMO and delay this fix which is needed to resolve CI (and customer) issues.

I understand that using pmem as the primary source of affinity information expands the scope of the work and I won't insist on including it for the initial pass at solving the problem. That having been said, the current approach will almost certainly need to be radically reworked in order to accommodate the new requirements. What I would rather see is a solution that addresses the problem now but in a way that will gracefully handle the future requirements. There are a lot of changes sprinkled around the codebase when it seems that this could be largely centralized in a single function in e.g. `server_utils.go`. It also sounds like the `first_core` config parameter is going to be ignored by the engine anyhow, so we should probably just go ahead and address that now, or at least make it easy to remove it later without too much churn.

I don't want to overengineer, but in my opinion the approach being proposed in this PR is adding technical debt. With a bit more work to incorporate context, I think this could solve the short and long-term problems.

**Comment by tanabarr** [comment]
> > @mjmac I would like to land this PR separately to the changes requested regarding using pmem as the canonical source of engine affinity. This PR specifically addresses a problem where the engine is started bound to a NUMA node that has no hugepages allocated on and DMA buffer memory allocation fails as a result. I think the changes to use pmem to derive affinity should be done as a separate PR as they will require nontrivial work to retrieve pmem socket-ID from scan results and integrate into storage config. That work would cloud the scope of this PR IMO and delay this fix which is needed to resolve CI (and customer) issues.
> 
> I understand that using pmem as the primary source of affinity information expands the scope of the work and I won't insist on including it for the initial pass at solving the problem. That having been said, the current approach will almost certainly need to be radically reworked in order to accommodate the new requirements. What I would rather see is a solution that addresses the problem now but in a way that will gracefully handle the future requirements. There are a lot of changes sprinkled around the codebase when it seems that this could be largely centralized in a single function in e.g. `server_utils.go`. It also sounds like the `first_core` config parameter is going to be ignored by the engine anyhow, so we should probably just go ahead and address that now, or at least make it easy to remove it later without too much churn.
> 
> I don't want to overengineer, but in my opinion the approach being proposed in this PR is adding technical debt. With a bit more work to incorporate context, I think this could solve the short and long-term problems.

Can I request clarity on the changes requested please. is this summary accurate:
- Move changes into a centralised function in server_utils.go

What do you require regarding first_core? This is not ignored by the engine. I don't really understand the comment. It is still used in a number of cases but just not the same time as pinned_numa_node.

I don't really understand why we can't land this and iterate on it, it solves the problem that we are facing.

**Comment by mjmac** [comment]
Will defer to @knard-intel and @kjacque for reviews on this, as a lot of the code came from my proposed updates. It looks good to me!

File: src/control/server/server_utils_test.go
Comment (kjacque) [inline src/control/server/server_utils_test.go:475]: Minor, but I suggest setting the affinities manually in this test instead of calling SetEngineAffinities. The SetEngineAffinities logic is nontrivial, and could complicate debugging the test.

**Review by kjacque — APPROVED** [review]
Comments are minor, could be addressed in a follow-on.

File: src/control/server/config/server_test.go
Comment (mjmac) [reply src/control/server/config/server_test.go:1399]: Yeah, this is a good point. I added this test before I'd refactored to pull responsibility for affinity detection out of this package. I was thinking that we might want to add more test cases to handle pmem affinity here, but I think you're right that it would be best to keep this package agnostic about those details and just make the test affinity sources completely generic. Unless you feel strongly though, I'd suggest doing this in the follow-on where the code is added to detect pmem affinity.

File: src/control/server/server_utils_test.go
Comment (mjmac) [reply src/control/server/server_utils_test.go:475]: Agreed. That having been said, I think the `Validate()` method(s) still have too many side-effects and should be refactored to make them read-only validations. This test relies on those side-effects and should be updated to be more explicit about the inputs. I don't think that necessarily needs to be part of this PR, though.

Comment (kjacque) [reply src/control/server/server_utils_test.go:475]: Agreed on all counts.

File: src/control/server/config/server.go
Comment (knard-intel) [inline src/control/server/config/server.go:683]: Does it imply that it will not be tolerated any more to have pinned_numa_node and first_core in the same engine section ?
If yes, it could be an issue as I have seen both settings used for example with the DAOS servers of the shared cluster.
However, I could probably miss something as I am not familiar with this portion of the code.

File: src/control/server/config/server.go
Comment (knard-intel) [inline src/control/server/config/server.go:706]: From what I understand this tests is not needed any more with test at line 683 ?

File: src/control/server/config/server.go
Comment (knard-intel) [inline src/control/server/config/server.go:696]: Seems a little odd for me to take the reference of a parameter variable (i.e. stack allocation). However, it is probably GO compliant.

**Review by knard-intel — COMMENTED** [review]
Mostly OK for me.
Just some points which have to be clarify: I probably miss something.

File: src/control/server/config/server.go
Comment (tanabarr) [reply src/control/server/config/server.go:706]: This clause is needed to support the case when `pinned == nil && core != 0`, what happens in that case is pinned gets set on line 696 and then needs to be unset here to avoid clashing. This was an interesting artifact of @mjmac which while it feels a bit strange is probably net positive due to making logic more readable IMO.

File: src/control/server/config/server.go
Comment (knard-intel) [reply src/control/server/config/server.go:683]: Good to know.

File: src/control/server/config/server.go
Comment (knard-intel) [reply src/control/server/config/server.go:706]: I miss this case. Thanks for the explanation.

**Review by knard-intel — APPROVED** [review]
Fine for me. Thanks for the clarification.

**Pull request merged**
